### PR TITLE
[CELEBORN-1156][BUILD] SBT publish support

### PR DESCRIPTION
**Pull request by cfmcgrady**
### What changes were proposed in this pull request?

As title

### Why are the changes needed?

As title

### Does this PR introduce _any_ user-facing change?

Yes, the user can publish shade clients via SBT

### How was this patch tested?

```shell
docker run -d -p 8081:8081 sonatype/nexus3
```

```shell
export SONATYPE_SNAPSHOTS_URL=http://192.168.3.46:8081/repository/maven-snapshots/
export SONATYPE_RELEASES_URL=http://192.168.3.46:8081/repository/maven-releases/
export ASF_USERNAME=admin
export ASF_PASSWORD=123456
```

- Publish the shade client for Spark 3.5:
```shell
./build/sbt -Pspark-3.4 celeborn-client-spark-3-shaded/publish
```

截屏2023-12-08 下午10 22 07

- Publish the shade client for Flink 1.18:

```shell
$ ./build/sbt -Pflink-1.18 celeborn-client-flink-1_18-shaded/publish
```
截屏2023-12-08 下午10 25 28

- Publish the shade client for MapReduce:
```shell
$ ./build/sbt -Pmr celeborn-client-mr-shaded/publish
```
截屏2023-12-08 下午10 25 47

### CODE REVIEW

File: docs/developers/sbt.md [hunk 1]
```diff
@@ -279,3 +279,43 @@ Similarly, if your objective involves compiling and packaging within an intranet
...
+
+| Environment Variable   | Description                                                                                                                      |
+| ---------------------- | -------------------------------------------------------------------------------------------------------------------------------- |
+| SONATYPE_HOST          | Sonatype repository host address, default is "oss.sonatype.org"                                                                  |
+| SONATYPE_USERNAME      | Sonatype repository username                                                                                                     |
+| SONATYPE_PASSWORD      | Sonatype repository password                                                                                                     |
+| SONATYPE_SNAPSHOTS_URL | Sonatype repository URL for snapshot version releases, default is "https://oss.sonatype.org/content/repositories/snapshots"      |
+| SONATYPE_RELEASES_URL  | Sonatype repository URL for official release versions, default is "https://oss.sonatype.org/service/local/staging/deploy/maven2" |
```
Comment (pan3793) [inline docs/developers/sbt.md:295]: Are SONATYPE_HOST redundant? I suppose SONATYPE_RELEASES_URL and SONATYPE_SNAPSHOTS_URL are sufficient.

For Apache release, the default URL should be https://repository.apache.org/service/local/staging/deploy/maven2 and https://repository.apache.org/content/repositories/snapshots

Comment (pan3793) [reply docs/developers/sbt.md:295]: we can eliminate SONATYPE_USERNAME, SONATYPE_PASSWORD and simply pick up ASF_USERNAME, ASF_PASSWORD, as they were used in source/binary release process

Comment (cfmcgrady) [reply docs/developers/sbt.md:295]: > For Apache release, the default URL should be https://repository.apache.org/service/local/staging/deploy/maven2 and https://repository.apache.org/content/repositories/snapshots

do you mean the default values(SONATYPE_SNAPSHOTS_URL/SONATYPE_RELEASES_URL) should be Apache URLs?

Comment (pan3793) [reply docs/developers/sbt.md:295]: Yea, we did such thing in another apache projects

Comment (cfmcgrady) [reply docs/developers/sbt.md]: > we can eliminate SONATYPE_USERNAME, SONATYPE_PASSWORD and simply pick up ASF_USERNAME, ASF_PASSWORD, as they were used in source/binary release process

done

File: project/CelebornBuild.scala [hunk 2]
```diff
@@ -218,6 +218,44 @@ object CelebornCommonSettings {
...
+    publishTo := {
+      if (isSnapshot.value) {
+        val publishUrl = sys.env.getOrElse("SONATYPE_SNAPSHOTS_URL", "https://repository.apache.org/content/repositories/snapshots")
+        Some(("snapshots" at publishUrl).withAllowInsecureProtocol(true))
+      } else {
+        val publishUrl = sys.env.getOrElse("SONATYPE_RELEASES_URL", "https://repository.apache.org/service/local/staging/deploy/maven2")
+        Some(("releases" at publishUrl).withAllowInsecureProtocol(true))
+      }
+    },
+    licenses += ("Apache-2.0", url("http://www.apache.org/licenses/LICENSE-2.0")),
```
Comment (pan3793) [inline project/CelebornBuild.scala:250]:
```suggestion
    licenses += ("Apache-2.0", url("https://www.apache.org/licenses/LICENSE-2.0")),
```

File: project/build.properties [hunk 3]
```diff
@@ -14,4 +14,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-sbt.version=1.9.4
+sbt.version=1.9.7
```
Comment (pan3793) [inline project/build.properties:17]: is it necessary? prefer to follow spark

File: docs/developers/sbt.md [hunk 4]
```diff
@@ -279,3 +279,41 @@ Similarly, if your objective involves compiling and packaging within an intranet
...
+| ---------------------- | ------------------------------------------------------------------------------------------------------------------------------------- |
+| ASF_USERNAME           | Sonatype repository username                                                                                                          |
+| ASF_PASSWORD           | Sonatype repository password                                                                                                          |
+| SONATYPE_SNAPSHOTS_URL | Sonatype repository URL for snapshot version releases, default is "https://repository.apache.org/content/repositories/snapshots"      |
+| SONATYPE_RELEASES_URL  | Sonatype repository URL for official release versions, default is "https://repository.apache.org/service/local/staging/deploy/maven2" |
+
+For example:
+```shell
+export SONATYPE_SNAPSHOTS_URL=http://192.168.3.46:8081/repository/maven-snapshots/
+export SONATYPE_RELEASES_URL=http://192.168.3.46:8081/repository/maven-releases/
```
Comment (pan3793) [inline docs/developers/sbt.md:299]: does the tailing slash matter?

File: project/build.properties (hunk 3 above)
Comment (cfmcgrady) [reply project/build.properties]: reverted.

File: docs/developers/sbt.md (hunk 4 above)
Comment (cfmcgrady) [reply docs/developers/sbt.md:299]: I've confirmed in a local environment that the presence or absence of a trailing slash doesn't affect the outcome.

**Review by pan3793 — APPROVED** [review]

**Comment by cfmcgrady** [comment]
thanks, merging to main(v0.4.0).

**Pull request closed without merging**